### PR TITLE
Wrong description for function ResolveEndpoint

### LINKED ISSUE #112246: kube-apiserver: ResolveEndpoint function is deterministic
### What happened?

When using Dynamic Admission Controllers such as a Mutating or a Validating Webhook configuration pointing to an in-cluster Service backed by multiple endpoints I noticed a reproducible distribution in the request load for a given amount of requests.

### What did you expect to happen?

When using a Webhook Configuration pointing to an in-cluster Service, the requests load should not be predictable. It should be forwarded to all the available endpoints in a non-deterministic way.

### How can we reproduce it (as minimally and precisely as possible)?

1. Create a Kubernetes cluster
2. Create a Deployment, serving an HTTPS endpoint that is returning an allowed Admission response, backed by 2 or more replicas; at every hit, print in the `stdout` the response
3. Expose the Deployment pods
4. Install a Mutating or a Validating Webhook Configuration, pointing it to the said Service 
5. Issue 100 HTTP requests that will be managed by the Deployment
6. Verify the hit on each Pod, it will be 33 for the first one, 33 for the second one, and 29 for the third.
7. Repeat multiple times, expecting the same results reported at point nr. 6

Using the `kubernetes` code base, the following snippet can be used to reproduce the faked random distribution.

```go
func TestResolve_Random(t *testing.T) {
	service := &v1.Service{
		ObjectMeta: metav1.ObjectMeta{
			Namespace: "default",
			Name:      "webhook",
		},
		Spec: v1.ServiceSpec{
			Type:      v1.ServiceTypeClusterIP,
			ClusterIP: "10.96.86.125",
			Ports: []v1.ServicePort{
				{
					Name:       "webhook",
					Port:       9443,
					TargetPort: intstr.FromInt(9443),
				},
			},
		},
	}

	serviceCache := cache.NewIndexer(cache.MetaNamespaceKeyFunc, cache.Indexers{cache.NamespaceIndex: cache.MetaNamespaceIndexFunc})
	serviceLister := v1listers.NewServiceLister(serviceCache)
	require.NoError(t, serviceCache.Add(service), "unexpected service add error")

	endpoints := []*v1.Endpoints{
		{
			ObjectMeta: metav1.ObjectMeta{
				Namespace: "default",
				Name:      "webhook",
			},
			Subsets: []v1.EndpointSubset{
				{
					Addresses: []v1.EndpointAddress{
						{
							IP: fuzzyEndpoint1,
						},
						{
							IP: fuzzyEndpoint2,
						},
						{
							IP: fuzzyEndpoint3,
						},
					},
					Ports: []v1.EndpointPort{
						{
							Name:     "webhook",
							Port:     9443,
							Protocol: v1.ProtocolTCP,
						},
					},
				},
			},
		},
	}

	endpointCache := cache.NewIndexer(cache.MetaNamespaceKeyFunc, cache.Indexers{cache.NamespaceIndex: cache.MetaNamespaceIndexFunc})
	endpointLister := v1listers.NewEndpointsLister(endpointCache)
	for _, ep := range endpoints {
		require.NoError(t, endpointCache.Add(ep), "unexpected endpoint add error")
	}

	testExecution := 100

	invocationCount := 100

	for te := 0; te < testExecution; te++ {
		invocationRes := map[string]int{
			fuzzyEndpoint1 + ":9443": 0,
			fuzzyEndpoint2 + ":9443": 0,
			fuzzyEndpoint3 + ":9443": 0,
		}

		for i := 0; i < invocationCount; i++ {
			u, err := ResolveEndpoint(serviceLister, endpointLister, service.Namespace, service.Name, 9443)
			require.NoError(t, err, "unexpected resolve error")

			if _, ok := invocationRes[u.Host]; !ok {
				t.Fatalf("received an IP that is not available in any Endpoint subset")
			}
			invocationRes[u.Host]++
		}
	}
}
```

### Anything else we need to know?

_No response_

### Kubernetes version

<details>

```console
$ kubectl version
Client Version: version.Info{Major:"1", Minor:"22", GitVersion:"v1.22.4", GitCommit:"b695d79d4f967c403a96986f1750a35eb75e75f1", GitTreeState:"clean", BuildDate:"2021-11-17T15:48:33Z", GoVersion:"go1.16.10", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"23", GitVersion:"v1.23.4", GitCommit:"e6c093d87ea4cbb530a7b2ae91e54c0842d8308a", GitTreeState:"clean", BuildDate:"2022-03-06T21:32:53Z", GoVersion:"go1.17.7", Compiler:"gc", Platform:"linux/amd64"}
```

</details>

### Cloud provider

<details>
N.R.
</details>

### OS version

<details>

N.R.

</details>

### Install tools

<details>

N.R.

</details>

### Container runtime (CRI) and version (if applicable)

<details>

N.R.

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

N.R.

</details>

### PR DESCRIPTION
**Pull request by prometherion**
#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:

Just a minor improvement due to a typo on the function `ResolveEndpoint` description.

#### Which issue(s) this PR fixes:

None

#### Special notes for your reviewer:

None

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

### CODE REVIEW

**Comment by prometherion** [comment]
Honestly, this is the quickest solution to address the issue, however, I would suggest a different approach.

https://github.com/kubernetes/kubernetes/blob/bcea98234f0fdc5ce4e976758977eec72b13bfd4/staging/src/k8s.io/apiserver/pkg/util/proxy/proxy.go#L68-L88

We could iterate over the subsets using the `range` form that shuffles by default the items of the slice, so we wouldn't need to replicate that behaviour using the `for` along with the random seed generation.

I can take care of providing another commit with the proposed solution, and eventually, squashing if it looks good to the reviewers.

**Comment by aojea** [comment]
we should have a test reproducing the issue for covering regressions

**Comment by prometherion** [comment]
As you suggested @aojea I'm working on writing a test to ensure no regression is introduced and I have to correct myself.

Actually, there's real randomization of the IP address selection, however, the result of the IP pick is reproducible, meaning I'm getting the exact distribution of the HTTP calls, from time to time.

To be correct, this is not entirely randomized, and the issue I reported is wrong in the form, thus I'm going to update it properly.

After these considerations, I will rework this PR to ensure we're using an accurate random address selection, and simplifying the code.

**Comment by prometherion** [comment]
The PR has been arranged according to the latest findings, and a simplified code has been proposed, looking forward to having a review and/or a discussion.

@aojea is still required a need for a regression test?

File: staging/src/k8s.io/apiserver/pkg/util/proxy/proxy.go [hunk 1]
```diff
@@ -64,28 +65,28 @@ func ResolveEndpoint(services listersv1.ServiceLister, endpoints listersv1.Endpo
...
-		for i := range ss.Ports {
-			if ss.Ports[i].Name == svcPort.Name {
-				// Pick a random address.
-				ip := ss.Addresses[rand.Intn(len(ss.Addresses))].IP
-				port := int(ss.Ports[i].Port)
+
+		for _, p := range ep.Ports {
+			if p.Name == svcPort.Name {
+				// Creating a new random generator to ensure non reproducibility
+				r := rand.New(rand.NewSource(time.Now().UnixNano()))
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/util/proxy/proxy.go:78]: a new random generator per request is unexpected

math.rand is seeded at process start in staging/src/k8s.io/component-base/cli/run.go

**Comment by liggitt** [comment]
/hold

I'm not seeing the problem with the existing code... starting iteration from a random endpoint and selecting a random address seems correct to me, even if the random sequence comes from math/rand.

File: staging/src/k8s.io/apiserver/pkg/util/proxy/proxy.go [hunk 2]
```diff
@@ -64,28 +65,28 @@ func ResolveEndpoint(services listersv1.ServiceLister, endpoints listersv1.Endpo
...
-	// Pick a random Subset to start searching from.
-	ssSeed := rand.Intn(len(eps.Subsets))
-
-	// Find a Subset that has the port.
-	for ssi := 0; ssi < len(eps.Subsets); ssi++ {
-		ss := &eps.Subsets[(ssSeed+ssi)%len(eps.Subsets)]
-		if len(ss.Addresses) == 0 {
+	// Looking for a random IP address that matches the Service port:
+	// Go natively shuffles slice items when iterating over it.
+	for _, ep := range eps.Subsets {
```
[the remark this answers is not in the document]
Comment (prometherion) [reply staging/src/k8s.io/apiserver/pkg/util/proxy/proxy.go:70]: You're right, I get confused with the `map` structure.

File: staging/src/k8s.io/apiserver/pkg/util/proxy/proxy.go (hunk 1 above)
Comment (prometherion) [reply staging/src/k8s.io/apiserver/pkg/util/proxy/proxy.go]: The proposed change has been removed.

**Comment by prometherion** [comment]
After the @liggit review, I dropped the commits proposing the change, except for the one regarding the typo for the function description.

**Comment by aojea** [comment]
/lgtm

The godoc is wrong and this fixes it
/assign @liggitt

**Comment by dims** [comment]
@liggitt please remove hold and add milestone for this to land in 1.26

**Comment by liggitt** [comment]
dropped the hold but this is fine to go into 1.27. I don't think we should add things to the merge queue that aren't improving tests or fixing bugs/blockers for 1.26

**Comment by dims** [comment]
based on discussion above, moving to 1.27

/milestone v1.27

**Pull request merged**